Reset noteheads when `StaveNote#setKeyLine` is called; Refactor threevoice_tests.js;

Silverwolf90 opened the pull request:
Take a look at the diffs. They are significant due to the new test setup affecting the automatic rest alignment.

Some of these tests relied on formatting code that was executed *only if* missing a stave (which really doesn't make sense to me why any kind of multi-voice formatting would be attempted in this case). Anyway, we are soon making this an invalid state. So now, the rests are formatted by code that formats StaveNotes with a stave. But the output is not good. The vertical adjustments will have to be completely rewritten to take real bounding box metrics into account (but this can only happen once the new stave requirement is in place), and changes will need to be made to support >2 voices.

See comment in code for more details.

0xfe commented:
Can you paste in some visual diffs please?

Silverwolf90 commented:
Ah yeah no problem, I'll make a post with those and walk through the changes too. Looking over the diffs again, I think some will need explanation.

Silverwolf90 commented on src/stavenote.js:
```diff
@@ -725,6 +739,7 @@ export class StaveNote extends StemmableNote {
   setKeyLine(index, line) {
     this.keyProps[index].line = line;
     this.note_heads[index].setLine(line);
+    this.reset();
```
This change is important. `StaveNote#reset()` rebuilds the `NoteHeads`, which clearly must be done after changing the line of a key. While this is a bug fix, it affects some of the test output in a negative way.

Silverwolf90 replied:
And actually the line above it probably needs to be removed (and `setLine` does nothing other than set a `line` property)

Silverwolf90 replied:
Also worth mentioning that while we could update the notehead individually, I think it's better to use the `.reset()` as a catch-all "something has changed, please rebuild" (which also indicates that formatting is happening before `.format` is called, but that's another story).

0xfe commented:
Also, at a high-level the three voice code is very fragile, and I'm not sure if we want to keep it this way. This is one of those cases where the benefit is not worth the cost, and we should probably look for a more general way to solve this problem.

Silverwolf90 commented:
First of all there are some (initially unexpected) diffs around rests in beams. These changes are related to [the line I added and described here.](https://github.com/0xfe/vexflow/pull/495/files#diff-127a29276b44875a1cde4a2ae983b9c9R742)

I recently found out that rests within beams are always aligned. But `Formatter#AlignRestsToNotes` uses `StaveNote#setKeyLine` which was broken until the above change in this PR. The rest wasn't shift position because the `NoteHead` (which contains the rest glyph) wasn't being updated. Now that it is being updated, we are seeing the following position changes (this is not all of them, but a sampling):

#### Auto_Beaming
![image](https://cloud.githubusercontent.com/assets/1631625/21395100/18cbb6ae-c793-11e6-852f-d40fa7d30da2.png)
![image](https://cloud.githubusercontent.com/assets/1631625/21395173/6f1629c2-c793-11e6-985b-e60d22f1c9b7.png)
![image](https://cloud.githubusercontent.com/assets/1631625/21395180/7322502c-c793-11e6-9fb6-2cc165eda3c6.png)
![auto_beaming beam_across_all_rests_with_stemlets_blessed](https://cloud.githubusercontent.com/assets/1631625/21395883/23871798-c796-11e6-824b-dc981691b8f2.png)
![auto_beaming beam_across_all_rests_with_stemlets_current](https://cloud.githubusercontent.com/assets/1631625/21395884/2389d03c-c796-11e6-8420-8447ae788a1a.png)
![auto_beaming beam_across_all_rests](https://cloud.githubusercontent.com/assets/1631625/21395885/238d2ade-c796-11e6-931d-2acf657f55bd.png)
![auto_beaming break_beams_on_middle_rests_only_blessed](https://cloud.githubusercontent.com/assets/1631625/21395886/239046ce-c796-11e6-816e-21bbc3abd4b5.png)
![auto_beaming break_beams_on_middle_rests_only_current](https://cloud.githubusercontent.com/assets/1631625/21395887/2392e6ae-c796-11e6-8159-a385094d845d.png)
![auto_beaming break_beams_on_middle_rests_only](https://cloud.githubusercontent.com/assets/1631625/21395888/23952176-c796-11e6-95e9-85e647be2c7c.png)

#### Tuplet.Complex_Tuplet
![image](https://cloud.githubusercontent.com/assets/1631625/21395467/7e994ffe-c794-11e6-87b5-cd4da7230f18.png)
![image](https://cloud.githubusercontent.com/assets/1631625/21395469/81508230-c794-11e6-9b2c-93662d96ef8c.png)
![image](https://cloud.githubusercontent.com/assets/1631625/21395476/843bbd70-c794-11e6-8647-51b9259878d7.png)

And here are the diffs for the tests I changed glyphs (remember that the tests with 3 voices s are now being formatted  with algorithm that wasn't designed for 3 voices):

![three_voice_rests auto_adjust_rest_positions___three_voices__1_blessed](https://cloud.githubusercontent.com/assets/1631625/21395817/e4d15b4e-c795-11e6-86dd-b670068a99c4.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__1_current](https://cloud.githubusercontent.com/assets/1631625/21395818/e4d442c8-c795-11e6-840e-918fc99c116b.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__1](https://cloud.githubusercontent.com/assets/1631625/21395819/e4d71534-c795-11e6-8301-23b53255459c.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__2_blessed](https://cloud.githubusercontent.com/assets/1631625/21395820/e4da512c-c795-11e6-895a-c22a2d63f5be.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__2_current](https://cloud.githubusercontent.com/assets/1631625/21395821/e4dd3ee6-c795-11e6-9fa5-dc7c7087c1ec.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__2](https://cloud.githubusercontent.com/assets/1631625/21395822/e4e4591a-c795-11e6-835b-b30c0d25ebfa.png)
![three_voice_rests auto_adjust_rest_positions___two_voices_blessed](https://cloud.githubusercontent.com/assets/1631625/21395823/e4e72e24-c795-11e6-887a-4c9140eb6e8f.png)
![three_voice_rests auto_adjust_rest_positions___two_voices_current](https://cloud.githubusercontent.com/assets/1631625/21395824/e4e94150-c795-11e6-88c6-1db7367479b2.png)
![three_voice_rests auto_adjust_rest_positions___two_voices](https://cloud.githubusercontent.com/assets/1631625/21395825/e4ed2eaa-c795-11e6-9ebc-f8dc1fb2d4ce.png)
![three_voice_rests three_voices____1_blessed](https://cloud.githubusercontent.com/assets/1631625/21395826/e4efcf8e-c795-11e6-8260-f299dc389f62.png)
![three_voice_rests three_voices____1_current](https://cloud.githubusercontent.com/assets/1631625/21395827/e4f721c6-c795-11e6-8b63-deb76e487d20.png)
![three_voice_rests three_voices____1](https://cloud.githubusercontent.com/assets/1631625/21395828/e4f7b320-c795-11e6-8d59-a5b38135c532.png)
![three_voice_rests three_voices____2_complex_blessed](https://cloud.githubusercontent.com/assets/1631625/21395831/e504900e-c795-11e6-8c07-b12259c8afd2.png)
![three_voice_rests three_voices____2_complex_current](https://cloud.githubusercontent.com/assets/1631625/21395829/e4fde4e8-c795-11e6-9439-2b62df903ab8.png)
![three_voice_rests three_voices____2_complex](https://cloud.githubusercontent.com/assets/1631625/21395830/e501c108-c795-11e6-91dc-e9ff221822b5.png)
![three_voice_rests three_voices____3_blessed](https://cloud.githubusercontent.com/assets/1631625/21395832/e5053cde-c795-11e6-9803-08167af32af4.png)
![three_voice_rests three_voices____3_current](https://cloud.githubusercontent.com/assets/1631625/21395833/e50c30ca-c795-11e6-99ad-755fd8140357.png)
![three_voice_rests three_voices____3](https://cloud.githubusercontent.com/assets/1631625/21395834/e510c194-c795-11e6-9929-cac47b4bc95b.png)

Silverwolf90 commented:
I added some padding to the displaced notes. Which improves some of the output. But concerning rests, those basically stay the same as the previous set of diffs. We'll need to rewrite `StaveNote.formatByY`.

Going to merge.

### New diffs
![accidental multi_voice_blessed](https://cloud.githubusercontent.com/assets/1631625/21508952/d64f6430-cc7c-11e6-9f46-974b4165d533.png)
![accidental multi_voice_current](https://cloud.githubusercontent.com/assets/1631625/21508953/d6507ed8-cc7c-11e6-8a2a-c47b72c13145.png)
![accidental multi_voice](https://cloud.githubusercontent.com/assets/1631625/21508954/d6540e54-cc7c-11e6-9a81-12fa4160c475.png)
![auto_beaming beam_across_all_rests_blessed](https://cloud.githubusercontent.com/assets/1631625/21508955/d6561d3e-cc7c-11e6-8776-e88118e3b752.png)
![auto_beaming beam_across_all_rests_current](https://cloud.githubusercontent.com/assets/1631625/21508956/d65929b6-cc7c-11e6-9734-34a6eac0b939.png)
![auto_beaming beam_across_all_rests_with_stemlets_blessed](https://cloud.githubusercontent.com/assets/1631625/21508957/d65af6f6-cc7c-11e6-85df-ebac9da1957e.png)
![auto_beaming beam_across_all_rests_with_stemlets_current](https://cloud.githubusercontent.com/assets/1631625/21508958/d6633870-cc7c-11e6-9b7e-500d602b5f24.png)
![auto_beaming beam_across_all_rests_with_stemlets](https://cloud.githubusercontent.com/assets/1631625/21508959/d6657946-cc7c-11e6-885a-37ca7921cec6.png)
![auto_beaming beam_across_all_rests](https://cloud.githubusercontent.com/assets/1631625/21508960/d668a6a2-cc7c-11e6-9b72-e0dcb3ce16dd.png)
![auto_beaming break_beams_on_middle_rests_only_blessed](https://cloud.githubusercontent.com/assets/1631625/21508961/d66d1606-cc7c-11e6-856d-20f46604f326.png)
![auto_beaming break_beams_on_middle_rests_only_current](https://cloud.githubusercontent.com/assets/1631625/21508962/d66dfc60-cc7c-11e6-971b-c3df42e95d75.png)
![auto_beaming break_beams_on_middle_rests_only](https://cloud.githubusercontent.com/assets/1631625/21508963/d67118a0-cc7c-11e6-9dce-a958a1bae052.png)
![dot multi_voice_blessed](https://cloud.githubusercontent.com/assets/1631625/21508964/d67616ca-cc7c-11e6-91ca-ad95dc255a44.png)
![dot multi_voice_current](https://cloud.githubusercontent.com/assets/1631625/21508965/d67947be-cc7c-11e6-9306-d1924d45a934.png)
![dot multi_voice](https://cloud.githubusercontent.com/assets/1631625/21508966/d67dbf7e-cc7c-11e6-9517-27523199b5b6.png)
![formatter stavenote_justification_blessed](https://cloud.githubusercontent.com/assets/1631625/21508967/d6821a92-cc7c-11e6-96a5-a55150893317.png)
![formatter stavenote_justification_current](https://cloud.githubusercontent.com/assets/1631625/21508968/d683d210-cc7c-11e6-8ca6-333261f3f531.png)
![formatter stavenote_justification](https://cloud.githubusercontent.com/assets/1631625/21508969/d68778a2-cc7c-11e6-9a2b-5e70503bebbc.png)
![rests auto_align_rests___multi_voice_blessed](https://cloud.githubusercontent.com/assets/1631625/21508970/d68bf558-cc7c-11e6-8212-e2522ef7f686.png)
![rests auto_align_rests___multi_voice_current](https://cloud.githubusercontent.com/assets/1631625/21508971/d68e5a78-cc7c-11e6-86d4-255305927ea9.png)
![rests auto_align_rests___multi_voice](https://cloud.githubusercontent.com/assets/1631625/21508972/d690549a-cc7c-11e6-92a5-48df991581ce.png)
![stavenote center_aligned_note___multi_voice_blessed](https://cloud.githubusercontent.com/assets/1631625/21508973/d6982c56-cc7c-11e6-927b-408bad9eaa86.png)
![stavenote center_aligned_note___multi_voice_current](https://cloud.githubusercontent.com/assets/1631625/21508974/d69b53d6-cc7c-11e6-90ad-3a3efc203fb4.png)
![stavenote center_aligned_note___multi_voice](https://cloud.githubusercontent.com/assets/1631625/21508975/d69de556-cc7c-11e6-88d1-0951ba6179fd.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__1_blessed](https://cloud.githubusercontent.com/assets/1631625/21508976/d6a06592-cc7c-11e6-87bf-121618320c0a.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__1_current](https://cloud.githubusercontent.com/assets/1631625/21508977/d6a40378-cc7c-11e6-890d-e474681981b5.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__1](https://cloud.githubusercontent.com/assets/1631625/21508978/d6ac326e-cc7c-11e6-9bd4-46773090003d.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__2_blessed](https://cloud.githubusercontent.com/assets/1631625/21508979/d6b23dee-cc7c-11e6-988c-aae9007a8517.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__2_current](https://cloud.githubusercontent.com/assets/1631625/21508980/d6b520ea-cc7c-11e6-9845-e668a5ff4bb8.png)
![three_voice_rests auto_adjust_rest_positions___three_voices__2](https://cloud.githubusercontent.com/assets/1631625/21508981/d6b81ab6-cc7c-11e6-9a9f-90e44cbe2c71.png)
![three_voice_rests auto_adjust_rest_positions___two_voices_blessed](https://cloud.githubusercontent.com/assets/1631625/21508982/d6bb965a-cc7c-11e6-9106-18e8dbfdab9b.png)
![three_voice_rests auto_adjust_rest_positions___two_voices_current](https://cloud.githubusercontent.com/assets/1631625/21508983/d6c22394-cc7c-11e6-8d56-1938ed2776d6.png)
![three_voice_rests auto_adjust_rest_positions___two_voices](https://cloud.githubusercontent.com/assets/1631625/21508984/d6c5966e-cc7c-11e6-8bcd-4531ed4bf590.png)
![three_voice_rests three_voices____1_blessed](https://cloud.githubusercontent.com/assets/1631625/21508985/d6c80dd6-cc7c-11e6-811d-3b2867d2573a.png)
![three_voice_rests three_voices____1_current](https://cloud.githubusercontent.com/assets/1631625/21508986/d6cad73c-cc7c-11e6-97bb-f2cbc5ea94e1.png)
![three_voice_rests three_voices____1](https://cloud.githubusercontent.com/assets/1631625/21508987/d6cfb70c-cc7c-11e6-9755-e99bf35c10d7.png)
![three_voice_rests three_voices____2_complex_blessed](https://cloud.githubusercontent.com/assets/1631625/21508988/d6d0955a-cc7c-11e6-9a19-9c330190f826.png)
![three_voice_rests three_voices____2_complex_current](https://cloud.githubusercontent.com/assets/1631625/21508989/d6d4caa8-cc7c-11e6-81fe-98e3233119c5.png)
![three_voice_rests three_voices____2_complex](https://cloud.githubusercontent.com/assets/1631625/21508991/d6d91568-cc7c-11e6-9588-40a45010e514.png)
![three_voice_rests three_voices____3_blessed](https://cloud.githubusercontent.com/assets/1631625/21508992/d6dbbf2a-cc7c-11e6-823f-62d4fb715102.png)
![three_voice_rests three_voices____3_current](https://cloud.githubusercontent.com/assets/1631625/21508993/d6de0f8c-cc7c-11e6-9c38-fbbc05ea449e.png)
![three_voice_rests three_voices____3](https://cloud.githubusercontent.com/assets/1631625/21508994/d6f48a32-cc7c-11e6-8b49-e094f0e9a367.png)
![tuplet complex_tuplet_blessed](https://cloud.githubusercontent.com/assets/1631625/21508995/d6f75aaa-cc7c-11e6-9dd3-2d4ded067820.png)
![tuplet complex_tuplet_current](https://cloud.githubusercontent.com/assets/1631625/21508996/d6fa0782-cc7c-11e6-849e-12a90708111f.png)
![tuplet complex_tuplet](https://cloud.githubusercontent.com/assets/1631625/21508997/d6fca30c-cc7c-11e6-995a-ea1e0a976afb.png)
